src/tree.rs

```diff
@@ -769,19 +640,326 @@ enum BuilderParentBy {
     KnownItem(Index),
 }
 
+/// Resolves the parent for a builder entry.
+struct ResolveParent<'a> {
+    builder: &'a Builder,
+    entry: &'a BuilderEntry,
+    problems: &'a mut Problems,
```

Comment:
A little surprised that this having the same lifetime as the non-mut things doesn't lead to problems, but I guess our use is limited.